litestar/_openapi/schema_generation/schema.py

```diff
@@ -549,14 +558,99 @@ def create_schema_for_pydantic_model(
         },
         type=OpenAPIType.OBJECT,
         title=_get_type_schema_name(field_type),
-        description=getdoc(field_type) or None,
+    )
+
+
+def create_schema_for_attrs_class(
+    field_type: type[AttrsInstance],
+    generate_examples: bool,
+    plugins: list[OpenAPISchemaPluginProtocol],
+    schemas: dict[str, Schema],
+) -> Schema:
+    """Create a schema object for a given attrs class.
+
+    Args:
+        field_type: An attrs class.
+        generate_examples: Whether to generate examples if none are given.
+        plugins: A list of plugins.
+        schemas: A mapping of namespaces to schemas - this mapping is used in the OA components section.
+
+    Returns:
+        A schema instance.
+    """
+    from attr import NOTHING
+    from attrs import fields_dict
+
+    field_type_hints = get_type_hints(field_type)
+    return Schema(
+        required=sorted(
+            [
+                field_name
+                for field_name, attribute in fields_dict(field_type).items()
+                if attribute.default is NOTHING and not is_optional_union(field_type_hints[field_name])
+            ]
+        ),
+        properties={
+            k: create_schema(
+                field=SignatureField.create(field_type=v, name=k),
+                generate_examples=generate_examples,
+                plugins=plugins,
+                schemas=schemas,
+            )
+            for k, v in field_type_hints.items()
+        },
+        type=OpenAPIType.OBJECT,
+        title=_get_type_schema_name(field_type),
+    )
+
+
+def create_schema_for_struct_class(
+    field_type: type[Struct],
+    generate_examples: bool,
+    plugins: list[OpenAPISchemaPluginProtocol],
+    schemas: dict[str, Schema],
+) -> Schema:
+    """Create a schema object for a given msgspec.Struct class.
+
+    Args:
+        field_type: A msgspec.Struct class.
+        generate_examples: Whether to generate examples if none are given.
+        plugins: A list of plugins.
+        schemas: A mapping of namespaces to schemas - this mapping is used in the OA components section.
+
+    Returns:
+        A schema instance.
+    """
+
+    field_type_hints = get_type_hints(field_type)
+    return Schema(
+        required=sorted(
+            [
+                struct.name
+                for struct in msgspec_struct_fields(field_type)
+                if struct.default is NODEFAULT
+                and struct.default_factory is NODEFAULT
+                and not is_optional_union(field_type_hints[struct.name])
+            ]
+        ),
+        properties={
+            k: create_schema(
+                field=SignatureField.create(field_type=v, name=k),
+                generate_examples=generate_examples,
+                plugins=plugins,
+                schemas=schemas,
+            )
+            for k, v in field_type_hints.items()
+        },
```

Comment:
This will exclude any [constraints](https://jcristharif.com/msgspec/constraints.html) set on types as interpreted by msgspec. Perhaps that's fine for your use case.

Note that we do provide our own JSON schema generation utility (https://jcristharif.com/msgspec/jsonschema.html) that properly handles all the types msgspec supports (including attrs classes) so the generated docs match the behavior. No pressure to use it, just letting you know it exists.

Reply:
Thanks, I'm aware. But we want consistent schema's - that's also why we don't use the pydantic stuff.